### PR TITLE
[Merged by Bors] - chore(IntegralClosure, Algebraic): renames, noncommutative generalizations and golfs

### PR DESCRIPTION
**Pull request by alreadydone**
[Zulip](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.238406.20IsIntegral.20dot.20notation.2C.20noncomm.20generalization.20.26.20golf/near/402028285)

Initially I just wanted to add more dot notations for IsIntegral and IsAlgebraic (done in #8437); then I noticed near-duplicates 
[Algebra.isIntegral_of_finite](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/IntegralClosure.html#Algebra.isIntegral_of_finite) `[Field R] [Ring A]` and 
[RingHom.IsIntegral.of_finite](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/IntegralClosure.html#RingHom.IsIntegral.of_finite) `[CommRing R] [CommRing A]`
so I went on to generalize the latter to cover the former, and generalized everything in the IntegralClosure file to the noncommutative case whenever possible.

In the process I noticed more golfs, which result in this PR. Most notably, [isIntegral_of_mem_of_FG](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/IntegralClosure.html#isIntegral_of_mem_of_FG) is now proven using [Cayley-Hamilton](https://math.stackexchange.com/a/2494433/12932) and doesn't depend on the Noetherian case [isIntegral_of_noetherian](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/IntegralClosure.html#isIntegral_of_noetherian); the latter is now proven using the former. In total the golfs makes mathlib 227 lines leaner (+487 -714).

The main changes are in the single file [RingTheory/IntegralClosure](https://github.com/leanprover-community/mathlib4/pull/8406/files#diff-426d60d7e9edc9e20379127612410ad633864c2d5a9cd001729d128f0993b00e):

+ Change the definition of `Algebra.IsIntegral` which makes it unfold to `IsIntegral` rather than `RingHom.IsIntegralElem` because the former has much more APIs.

+ Fix lemma names involving `is_integral` which are actually about `IsIntegralElem`:
`RingHom.is_integral_map` → `RingHom.isIntegralElem_map`
`RingHom.is_integral_of_mem_closure` → `RingHom.IsIntegralElem.of_mem_closure`
`RingHom.is_integral_zero/one` → `RingHom.isIntegralElem_zero/one`
`RingHom.is_integral_add/neg/sub/mul/of_mul_unit` → `RingHom.IsIntegralElem.add/neg/sub/mul/of_mul_unit`

+ Add a lemma `Algebra.IsIntegral.of_injective`.

+ Move `isIntegral_of_(submodule_)noetherian` down and golf them.

+ Remove `(Algebra.)isIntegral_of_finite` that work only over fields, in favor of the more general `(Algebra.)isIntegral.of_finite`.

+ Merge duplicate lemmas `isIntegral_of_isScalarTower` and `isIntegral_tower_top_of_isIntegral` into `IsIntegral.tower_top`.

+ Golf `IsIntegral.of_mem_of_fg` by first proving `IsIntegral.of_finite` using Cayley-Hamilton.

+ Add a docstring mentioning the Kurosh problem at `Algebra.IsIntegral.finite`. The negative solution to the problem means the theorem doesn't generalize to noncommutative algebras.

+ Golf `IsIntegral.tmul` and `isField_of_isIntegral_of_isField(')`.

+ Combine `isIntegral_trans_aux` into `isIntegral_trans` and golf.

+ Add `Algebra` namespace to `isIntegral_sup`.

+ rename lemmas for dot notation:
`RingHom.isIntegral_trans` → `RingHom.IsIntegral.trans`
`RingHom.isIntegral_quotient/tower_bot/top_of_isIntegral` → `RingHom.IsIntegral.quotient/tower_bot/top`
`isIntegral_of_mem_closure'` → `IsIntegral.of_mem_closure'` (and the '' version)
`isIntegral_of_surjective` → `Algebra.isIntegral_of_surjective`

The next changed file is RingTheory/Algebraic:

+ Rename:
`of_larger_base` → `tower_top` (for consistency with `IsIntegral`)
`Algebra.isAlgebraic_of_finite` → `Algebra.IsAlgebraic.of_finite`
`Algebra.isAlgebraic_trans` → `Algebra.IsAlgebraic.trans`

+ Add new lemmas`Algebra.IsIntegral.isAlgebraic`, `isAlgebraic_algHom_iff`, and `Algebra.IsAlgebraic.of_injective` to streamline some proofs.

+ Add aliases `(Algebra.)IsAlgebraic.IsIntegral` (in the field case).

The generalization from CommRing to Ring requires an additional lemma `scaleRoots_eval₂_mul_of_commute` in Polynomial/ScaleRoots.

A lemma `Algebra.lmul_injective` is added to Algebra/Bilinear (in order to golf the proof of `IsIntegral.of_mem_of_fg`).

In all other files, I merely fix the changed names, or use newly available dot notations.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/IntegralClosure.lean [hunk 1]
```diff
@@ -973,135 +819,102 @@ open Algebra
...
+    convert hp; apply p.map_toSubring S.toSubring⟩
+  let Sx := Subalgebra.toSubmodule (adjoin S {x})
+  let MSx : Module S Sx := SMulMemClass.toModule _ -- the next line times out without this
+  have : Module.Finite S Sx := ⟨(Submodule.fg_top _).mpr hSx.fg_adjoin_singleton⟩
+  refine .of_mem_of_fg ((adjoin S {x}).restrictScalars R) ?_ _
+    ((Subalgebra.mem_restrictScalars R).mpr <| subset_adjoin rfl)
+  rw [← Submodule.fg_top, ← Module.finite_def]
+  letI : SMul S Sx := { MSx with } -- need this even though MSx is there
+  have : IsScalarTower R S Sx :=
+    Submodule.isScalarTower Sx -- Lean looks for `Module A Sx` without this
```
Comment (alreadydone) [inline Mathlib/RingTheory/IntegralClosure.lean:848]: Some typeclass inference failures that may be worth investigating (but are probably well known)

**Comment by alreadydone** [comment]
cc:
original author of [isIntegral_of_mem_of_FG](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/IntegralClosure.html#isIntegral_of_mem_of_FG) @kckennylau
@erdOne who introduced the generalized Cayley-Hamilton [LinearMap.exists_monic_and_aeval_eq_zero](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Matrix/Charpoly/LinearMap.html#LinearMap.exists_monic_and_aeval_eq_zero) allowing to golf the above 97-line proof down to 1+2 lines (I guess it wasn't golfed due to Module.End being noncommutative?)
original author of [isField_of_isIntegral_of_isField](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/IntegralClosure.html#isField_of_isIntegral_of_isField) @dtumad: I'm indeed able to use the `Polynomial.reverse` and `Polynomial.DivX` API
original author of [isIntegral_trans_aux](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/IntegralClosure.html#isIntegral_trans_aux) @jcommelin (the aux lemma doesn't allow noncommutative B, so it's now removed, and the main proof now adjoins elements in A rather than in B)
@eric-wieser @haruhisa-enomoto who may be interested in the noncommutative generalization

**Comment by eric-wieser** [comment]
Is it possible to split the dot notation changes to their own PR? It would be much better if the mathematical and stylistic content were in separate reviews.

File: Mathlib/RingTheory/IntegralClosure.lean [hunk 2]
```diff
@@ -66,133 +66,82 @@ variable (A)
...
-#align algebra.is_integral_of_finite Algebra.isIntegral_of_finite
-
-end
-
-section
-
 variable {R A B S : Type*}
 
-variable [CommRing R] [CommRing A] [CommRing B] [CommRing S]
+variable [CommRing R] [CommRing A] [Ring B] [CommRing S]
```
Comment (alreadydone) [inline Mathlib/RingTheory/IntegralClosure.lean:88]: The only change to lemma statements is here.

**Comment by alreadydone** [comment]
> Is it possible to split the dot notation changes to their own PR? It would be much better if the mathematical and stylistic content were in separate reviews.

As I said on Zulip:

> The PR golfs some long proofs which collectively trims down mathlib4 by 226 lines, but only adds two new lemmas, removes an auxiliary lemma, and deduplicates some lemmas, so it's hopefully not hard to review.
> The PR can't be easily split, as the renamed/reordered lemmas are also used in the golfs.

There is not much "mathematical content" because all I did to the lemma statements is [here](https://github.com/leanprover-community/mathlib4/pull/8406#discussion_r1393601820) (replacing [CommRing] by [Ring]). Lemmas may be moved (reordered), golfed, and deduplicated, but the statements are only changed (generalized) in this way, and these are all concentrated in the single file (plus one lemma in Polynomial/ScaleRoots and Algebra/Bilinear each).

On the other hand, the stylistic (dot notation) changes require fixing lots of files, but those fixes are trivial to review. If I first introduce dot notation in a separate PR, I'll need to fix long proofs (e.g. `isIntegral_of_mem_of_FG`) that will simply be completely removed in the subsequent golfing PR. If I first do the golfs in a separate PR, I'll need to use the old lemma names and switch to new lemma names in the subsequent dot notations PR. So both options don't seem very economic, though I would choose the second option if you insist on a split.

**Comment by jcommelin** [comment]
Thanks a lot for this PR. It's a nice clean-up.

I do wonder whether we can make more use of new Lean 4 features now. Since this PR is renaming a lot of things, it might be good to keep the old names around and use the `@[deprecated]` feature to direct users to the new name.

@semorrison Is there tooling to easily get a list of all decls that are being removed/added in this PR?

**Comment by semorrison** [comment]
> @semorrison Is there tooling to easily get a list of all decls that are being removed/added in this PR?

No, but @j-loreaux [explained](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.238389.20split.20MeasureSpace.2Elean.20into.203.20files/near/401851090) how to implemented this. I'll review! :-)

**Comment by alreadydone** [comment]
I think some fuzzy matches may be needed for this PR, as some lemmas changed both name and statement (CommRing -> Ring). 

Applying retroactively to #8358 may be easier, as statements are not changed there.

**Comment by eric-wieser** [comment]
> I think some fuzzy matches may be needed for this PR, as some lemmas changed both name and statement (CommRing -> Ring). Mathlib changelog probably can't trace through such changes.

This sounds like an argument for not mixing big renames with other refactors...

**Comment by j-loreaux** [comment]
@semorrison I'll work on it.

**Comment by erdOne** [comment]
I think `Algebra.IsIntegral` and `Algebra.IsAlgebraic` should be classes instead. Should we change it (in the future)? Will such a refactor reverse some changes made in this PR?

**Comment by erdOne** [comment]
Also, does it conflict with #7857 significantly?

**Comment by alreadydone** [comment]
> I think some fuzzy matches may be needed for this PR, as some lemmas changed both name and statement (CommRing -> Ring). Mathlib changelog probably can't trace through such changes.

On a second thought, in fact the #align commands already provide a way to track declarations renamed in mathlib4; people tend to modify the #aligns (possibly adding ₓ) but rarely remove them. Ideally the mathlib changelog could track all mathlib4 declaration names that have ever been aligned to a mathlib3 declaration and display them in the same page (is it the case now?). However, declarations that don't exist in mathlib3 don't have associated #align commands. Maybe we could introduce a #rename command and ask PR authors to add such a command whenever they rename a declaration (and the changelog could track renaming history transitively, which will be distinct from #aligning to a fixed base mathlib3 declaration), or maybe such commands can be auto-generated, and all the author has to do is to use a UI to pick which new declarations correspond to which old declarations. There may be auto-generated declarations that disappear or get added, but as a first approximation we may ignore them, as dealing with them may require building the whole mathlib rather than building a few modules and parsing some text.

Once we align renamed declarations, we can focus on comparing the declarations. GitHub offers a UI for `git diff` and we might want a `decl diff` UI as well. It should check the statement elaborates to the same thing, verify the arguments are the same (either strictly or up to reordering), highlight any the changed parts of the statement (e.g. generalized typeclasses) and definition body with red and green colors, and optionally allow to compare golfed proofs with the original. I imagine this will help reviews significantly, especially for auditing potential AI-assisted big refactors in the future (I wrote down some ideas on [Zulip](https://leanprover.zulipchat.com/#narrow/stream/219941-Machine-Learning-for-Theorem-Proving/topic/Morph.20Prover.20v0.207B/near/398564304)).

**Comment by j-loreaux** [comment]
@jcommelin The way I generated this isn't ready for prime time, but here is a list of declarations which were added/removed (+/-) in this PR:
```
+Algebra.IsAlgebraic.of_finite
+Algebra.IsAlgebraic.of_larger_base
+Algebra.IsAlgebraic.of_larger_base_of_injective
+Algebra.IsAlgebraic.trans
+Algebra.IsIntegral.quotient
+Algebra.IsIntegral.trans
-Algebra.isAlgebraic_of_finite
-Algebra.isAlgebraic_of_larger_base
-Algebra.isAlgebraic_of_larger_base_of_injective
-Algebra.isAlgebraic_trans
-Algebra.isIntegral_of_finite
+Algebra.isIntegral_of_surjective
+Algebra.isIntegral_sup
-Algebra.isIntegral_trans
+Algebra.lmul_injective
-FG_adjoin_of_finite
-FG_adjoin_singleton_of_integral
+IsAlgebraic.algHom
+IsAlgebraic.algebraMap
+IsAlgebraic.of_finite
+IsAlgebraic.of_larger_base
+IsAlgebraic.of_larger_base_of_injective
+IsAlgebraic.of_pow
+IsIntegral.add
+IsIntegral.fg_adjoin_singleton
+IsIntegral.map
+IsIntegral.map_of_comp_eq
+IsIntegral.mul
+IsIntegral.neg
+IsIntegral.of_finite
+IsIntegral.of_mem_closure
+IsIntegral.of_mem_closure'
+IsIntegral.of_mem_closure''
+IsIntegral.of_mem_of_fg
+IsIntegral.of_mul_unit
+IsIntegral.of_pow
+IsIntegral.of_subring
+IsIntegral.smul
+IsIntegral.sub
+IsIntegral.tower_bot
+IsIntegral.tower_bot_of_field
+IsIntegral.tower_top
+Polynomial.scaleRoots_eval₂_mul_of_commute
+RingHom.IsIntegral.quotient
+RingHom.IsIntegral.tower_bot
+RingHom.IsIntegral.tower_top
+RingHom.IsIntegral.trans
+RingHom.IsIntegralElem.add
+RingHom.IsIntegralElem.mul
+RingHom.IsIntegralElem.neg
+RingHom.IsIntegralElem.of_mem_closure
+RingHom.IsIntegralElem.of_mul_unit
+RingHom.IsIntegralElem.sub
+RingHom.isIntegralElem.of_comp
-RingHom.isIntegralElem_of_isIntegralElem_comp
-RingHom.isIntegral_of_isIntegral_mul_unit
+RingHom.isIntegralElem_map
+RingHom.isIntegralElem_one
+RingHom.isIntegralElem_zero
-RingHom.isIntegral_quotient_of_isIntegral
-RingHom.isIntegral_tower_bot_of_isIntegral
-RingHom.isIntegral_tower_top_of_isIntegral
-RingHom.isIntegral_trans
-RingHom.is_integral_add
-RingHom.is_integral_map
-RingHom.is_integral_mul
-RingHom.is_integral_neg
-RingHom.is_integral_of_mem_closure
-RingHom.is_integral_one
-RingHom.is_integral_sub
-RingHom.is_integral_zero
+fg_adjoin_of_finite
-isAlgebraic_algHom_of_isAlgebraic
-isAlgebraic_algebraMap_of_isAlgebraic
-isAlgebraic_of_finite
-isAlgebraic_of_larger_base
-isAlgebraic_of_larger_base_of_injective
-isAlgebraic_of_pow
-isIntegral_add
-isIntegral_map_of_comp_eq_of_isIntegral
-isIntegral_mul
-isIntegral_neg
-isIntegral_ofSubring
-isIntegral_of_finite
-isIntegral_of_isIntegral_mul_unit
-isIntegral_of_isScalarTower
-isIntegral_of_mem_closure
-isIntegral_of_mem_closure'
-isIntegral_of_mem_closure''
-isIntegral_of_mem_of_FG
-isIntegral_of_pow
-isIntegral_of_surjective
-isIntegral_quotient_of_isIntegral
-isIntegral_smul
-isIntegral_sub
-isIntegral_sup
-isIntegral_tower_bot_of_isIntegral
-isIntegral_tower_bot_of_isIntegral_field
-isIntegral_tower_top_of_isIntegral
-isIntegral_trans_aux
-map_isIntegral
-mem_integralClosure_iff_mem_FG
+mem_integralClosure_iff_mem_fg
```

**Comment by alreadydone** [comment]
@j-loreaux Thanks! Here's a paired up version by reordering your list:
```
The one removed lemma:
-isIntegral_trans_aux

Two new lemmas:
+Polynomial.scaleRoots_eval₂_mul_of_commute
+Algebra.lmul_injective


-Algebra.isAlgebraic_of_finite
+Algebra.IsAlgebraic.of_finite

-Algebra.isAlgebraic_of_larger_base
+Algebra.IsAlgebraic.of_larger_base

-Algebra.isAlgebraic_of_larger_base_of_injective
+Algebra.IsAlgebraic.of_larger_base_of_injective

-Algebra.isAlgebraic_trans
+Algebra.IsAlgebraic.trans

-isIntegral_quotient_of_isIntegral
+Algebra.IsIntegral.quotient

-Algebra.isIntegral_trans
+Algebra.IsIntegral.trans

-Algebra.isIntegral_of_finite
+Algebra.isIntegral_of_surjective

-isIntegral_sup
+Algebra.isIntegral_sup


-FG_adjoin_of_finite
+fg_adjoin_of_finite

-FG_adjoin_singleton_of_integral
+IsIntegral.fg_adjoin_singleton

-isAlgebraic_algHom_of_isAlgebraic
+IsAlgebraic.algHom

-isAlgebraic_algebraMap_of_isAlgebraic
+IsAlgebraic.algebraMap

-isAlgebraic_of_finite
+IsAlgebraic.of_finite

-isAlgebraic_of_larger_base
+IsAlgebraic.of_larger_base

-isAlgebraic_of_larger_base_of_injective
+IsAlgebraic.of_larger_base_of_injective

-isAlgebraic_of_pow
+IsAlgebraic.of_pow


-isIntegral_add
+IsIntegral.add

-isIntegral_map_of_comp_eq_of_isIntegral
+IsIntegral.map_of_comp_eq

-isIntegral_mul
+IsIntegral.mul

-isIntegral_neg
+IsIntegral.neg

-isIntegral_ofSubring
+IsIntegral.of_subring

-isIntegral_of_finite
+IsIntegral.of_finite

-isIntegral_of_isIntegral_mul_unit
+IsIntegral.of_mul_unit

-isIntegral_of_mem_closure
+IsIntegral.of_mem_closure

-isIntegral_of_mem_closure'
+IsIntegral.of_mem_closure'

-isIntegral_of_mem_closure''
+IsIntegral.of_mem_closure''

-isIntegral_of_mem_of_FG
+IsIntegral.of_mem_of_fg

-isIntegral_of_pow
+IsIntegral.of_pow

-isIntegral_of_surjective

-isIntegral_smul
+IsIntegral.smul

-isIntegral_sub
+IsIntegral.sub

-isIntegral_tower_bot_of_isIntegral
+IsIntegral.tower_bot

-isIntegral_tower_bot_of_isIntegral_field
+IsIntegral.tower_bot_of_field

-isIntegral_of_isScalarTower
-isIntegral_tower_top_of_isIntegral
+IsIntegral.tower_top

-map_isIntegral
+IsIntegral.map

-mem_integralClosure_iff_mem_FG
+mem_integralClosure_iff_mem_fg


-RingHom.isIntegral_quotient_of_isIntegral
+RingHom.IsIntegral.quotient

-RingHom.isIntegral_tower_bot_of_isIntegral
+RingHom.IsIntegral.tower_bot

-RingHom.isIntegral_tower_top_of_isIntegral
+RingHom.IsIntegral.tower_top

-RingHom.isIntegral_trans
+RingHom.IsIntegral.trans

-RingHom.is_integral_add
+RingHom.IsIntegralElem.add

-RingHom.is_integral_map
+RingHom.isIntegralElem_map

-RingHom.is_integral_mul
+RingHom.IsIntegralElem.mul

-RingHom.is_integral_neg
+RingHom.IsIntegralElem.neg

-RingHom.is_integral_of_mem_closure
+RingHom.IsIntegralElem.of_mem_closure

-RingHom.is_integral_one
+RingHom.isIntegralElem_one

-RingHom.is_integral_sub
+RingHom.IsIntegralElem.sub

-RingHom.is_integral_zero
+RingHom.isIntegralElem_zero

-RingHom.isIntegralElem_of_isIntegralElem_comp
+RingHom.isIntegralElem.of_comp

-RingHom.isIntegral_of_isIntegral_mul_unit
+RingHom.IsIntegralElem.of_mul_unit
```

If I knew about `@[deprecated]` (and that it's a good idea to use it here) I would have added them as I rename the lemmas. Since I didn't do that, should there be a utility that takes the list above and automatically add the @[deprecated ...] attributes? (Since most declarations are not in a `namespace` section, this should be easy for this PR.) Could one declaration have multiple `deprecated` attributes (I'd like that for `IsIntegral.tower_top`)?

**Comment by alreadydone** [comment]
> I do wonder whether we can make more use of new Lean 4 features now. Since this PR is renaming a lot of things, it might be good to keep the old names around and use the `@[deprecated]` feature to direct users to the new name.

@jcommelin Can you clarify how you intend to apply `@[deprecated]` attributes to this PR? From what I see in mathlib4, `@[deprecated X]` is added to a deprecated declaration Y to signify that X should be used instead of Y, so the old lemma Y need to be kept rather than renamed. The lemmas X and Y usually don't have exactly the same statement (for example [RingHom.map_list_prod](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/BigOperators/Basic.html#RingHom.map_list_prod) uses `RingHom` while [map_list_prod](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/List/BigOperators/Basic.html#map_list_prod) uses `MonoidHomClass`), so the attribute doesn't seem to be intended for renamed lemmas.

Of course I can make the new names aliases of the old ones, then this PR would be adding aliases rather than renaming (and I wouldn't even need to fix the 30+ files), but is such proliferation of aliases desirable? Every alias will appear in the docs in its full form, I think. Note that some old names are simply problematic (e.g. `isIntegral_sup` is lacking namespace) and should be renamed anyway.

**Comment by jcommelin** [comment]
I want to preface my comment by saying that this is my personal opinion. As a community, we're still very much trying to figure out what the best strategy/policy is, I guess.

Anyways, I think I would advocate keeping the old names around, and tagging them with `@[deprecated fooBar]` as a pointer to the new name. In a future PR (can be hours/days/weeks later) we remove all the deprecated decls. The big upshot is that there will be a formal trace in the git history that traces the renaming. This will help:

- people bumping mathlib in their personal projects
- changelog
- future automated refactoring tools that do mathlib bumps for us

**Comment by jcommelin** [comment]
@j-loreaux Thanks for generating that list, and @alreadydone thanks for pairing up the old and new names. Would you mind preserving that list in the git history by copying it to the top message?

**Comment by eric-wieser** [comment]
Since @adomani has gone to the trouble of reducing the diff, I think we should merge #8437 before this PR.

**Comment by alreadydone** [comment]
@adomani's #8437 did part of the renames, but the following were not included:

In IntegralClosure:
`RingHom.is_integral_map` → `RingHom.isIntegralElem_map`
`RingHom.is_integral_of_mem_closure` → `RingHom.IsIntegralElem.of_mem_closure`
`RingHom.is_integral_zero/one` → `RingHom.isIntegralElem_zero/one`
`RingHom.is_integral_add/neg/sub/mul/of_mul_unit` → `RingHom.IsIntegralElem.add/neg/sub/mul/of_mul_unit`
`RingHom.isIntegral_trans` → `RingHom.IsIntegral.trans`
`RingHom.isIntegral_quotient/tower_bot/top_of_isIntegral` → `RingHom.IsIntegral.quotient/tower_bot/top`
`isIntegral_sup` → `Algebra.isIntegral_sup`
`IsIntegral.of_mem_closure'` → `IsIntegral.of_mem_closure'`
`isIntegral_of_surjective` → `Algebra.isIntegral_of_surjective`

In RingTheory/Algebraic:
`of_larger_base` → `tower_top` (for consistency with `IsIntegral`)
`Algebra.isAlgebraic_of_finite` → `Algebra.IsAlgebraic.of_finite`
`Algebra.isAlgebraic_trans` → `Algebra.IsAlgebraic.trans`

I also added a lemma `Algebra.IsIntegral.of_injective` in IntegralClosure and three in RingTheory/Algebraic: `Algebra.IsIntegral.isAlgebraic`, `isAlgebraic_algHom_iff`, and `Algebra.IsAlgebraic.of_injective`.

(Currently I can't easily use @j-loreaux's #8433 to generate decl name changes, because if I merge this PR directly into #8433's branch the diff generated would include new commits on master. If I first merge master into #8433's branch, then I need to wait for the cache, and then merge this PR into #8433 and wait for the cache again, which is time consuming.)

I also decided to make R implicit in [IsIntegral.isAlgebraic](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Algebraic.html#IsIntegral.isAlgebraic), which resulted in more files needing to be fixed.

@adomani's PR means that the current changes in IntegralClosure and RingTheory/Algebraic only cause 24 files to break, so it's possible to split PR into one that changes 24 files and another one that changes 17 files (depending on the previous one, which would just be using the new dot notations to golf proofs), but I'm not sure it's worth doing; most diffs would still be in the first PR. If @adomani would do another renaming PR and include some changes in implicitness, then the number 24 would decrease and a split may be more attractive.

**Comment by jcommelin** [comment]
Can you please add the remaining renames to the PR commit message?

**Comment by jcommelin** [comment]
I'm inclined to merge this as is. Damiano's tool was a first success. And hopefully we can use it more in the future. But splitting up this PR seems like it will be a lot of work for little remaining gain.

**Comment by alreadydone** [comment]
Thanks @jcommelin! PR message edited.

**Comment by adomani** [comment]
The renames were an experiment to test an automated renaming tool.  Since it was the first "real world" attempt, I aimed to rename only lemmas for which there was no dot-notation involved.

Currently, the tool finds the positions where the old declaration appears and replaces them with the new one.  No fancy namespacing/dot-notation/removal of arguments is implemented.

I would like to add namespacing options, since that is easy to implement as a string-based replacement.

However, doing more context/syntax-aware changes is best done with a tool implemented in Lean.

The tool itself, reads the location of the syntax from the `ilean` files and then uses bash-commands to do the string replacements.

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Comment by j-loreaux** [comment]
@alreadydone yeah, #8433 is not usable until it's merged because it changes `lakefile.lean`. Sorry!

**Pull request closed without merging**